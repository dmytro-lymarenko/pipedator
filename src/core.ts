export interface ValidationError {
	message: string;
	path: string[]; // ['a', '0', 'test'] which is equivalent to 'a[0].test' of the value
}

export interface ValidationContext {
	value: any;
	path: string[];
}

export interface CreateValidatorOptions {
	validate: (value: any, ctx: ValidationContext) => null | ValidationError;
}

export interface Validator {
	validate: (value: any, ctx?: ValidationContext) => null | ValidationError;
	pipe: (...validators: Validator[]) => Validator;
}

export function pipe(...validators: Validator[]) {
	return createValidator({
		validate: (value, ctx) => {
			// find the first error in pipe
			let i: number = 0;
			let error: ValidationError | null = null;

			while (i < validators.length && error === null) {
				error = validators[i].validate(value, ctx);
				i = i + 1;
			}

			return error;
		},
	});
}

/**
 * Don't use ${value} inside message when value is invalid
 * @param options
 */
export function createValidator(options: CreateValidatorOptions): Validator {
	function validate(value: any, ctx?: ValidationContext) {
		return options.validate(value, ctx || { value, path: [] });
	}

	return {
		validate,
		pipe: (...validators) =>
			pipe(
				createValidator({ validate }),
				...validators
			),
	};
}
