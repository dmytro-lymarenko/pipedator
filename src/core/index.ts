export interface ValidationError {
	message: string;
	path: string;
	children: ValidationError[] | null;
}

export interface ValidationContext {
	value: any;
	path: string[];
	rootValue: any;
}

export interface CreateValidatorOptions {
	validate: (value: any, ctx: ValidationContext) => null | ValidationError;
}

export interface Validator<ValidValue = any> {
	validate: (value: any, ctx?: ValidationContext) => null | ValidationError;
	isValid: (value: any) => value is ValidValue;
}

export function findFirstError(
	getValidator: (index: number) => Validator,
	getValue: (index: number) => any,
	getContext: (index: number) => ValidationContext,
	length: number
): { error: ValidationError | null; index: number } {
	let i: number = 0;
	let error: ValidationError | null = null;

	while (i < length && error === null) {
		error = getValidator(i).validate(getValue(i), getContext(i));
		i = i + 1;
	}

	return { error, index: i - 1 };
}

export function getFirstErrors(
	getValidator: (index: number) => Validator,
	getValue: (index: number) => any,
	getContext: (index: number) => ValidationContext,
	length: number
): ValidationError[] {
	const errors: ValidationError[] = [];
	let i: number = 0;
	let error: ValidationError | null = null;

	do {
		error = getValidator(i).validate(getValue(i), getContext(i));
		if (error) {
			errors.push(error);
		}
		i = i + 1;
	} while (i < length && error !== null);

	return errors;
}

/**
 * Don't use ${value} inside message when value is invalid
 * @param options
 */
export function createValidator<ValidValue = any>(options: CreateValidatorOptions): Validator<ValidValue> {
	function validate(value: any, ctx?: ValidationContext) {
		const context: ValidationContext = ctx || {
			value,
			path: [],
			rootValue: value,
		};

		return options.validate(value, context);
	}

	function isValid(value: any): value is ValidValue {
		return validate(value) === null;
	}

	return {
		validate,
		isValid,
	};
}

export function validate(validator: Validator, value: any) {
	const ctx: ValidationContext = {
		value,
		path: [],
		rootValue: value,
	};

	return validator.validate(value, ctx);
}

export function getCurrentPath(ctx: ValidationContext): string {
	return ctx.path.length === 0 ? '' : ctx.path[ctx.path.length - 1];
}

export interface ValidationRef {
	(ctx: ValidationContext): any;
}

export function ref(path: string[]): ValidationRef {
	return (ctx: ValidationContext) => path.reduce((v, key) => v && v[key], ctx.rootValue);
}
