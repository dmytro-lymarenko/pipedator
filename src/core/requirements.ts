export interface BaseRequirement {
	value: any;
	path: string[];
	message: string;
}

export interface RequirementFactory<R extends Requirement = Requirement> {
	(path: string[], value: any): R;
}

export type Requirement = SingleRequirement | DependenceRequirement | AlternativeRequirement | ConcatenationRequirement;

export interface SingleRequirement extends BaseRequirement {
	type: 'single';
}

export interface DependenceRequirement extends BaseRequirement {
	type: 'dependence';
	requirement: Requirement;
}

export interface AlternativeRequirement extends BaseRequirement {
	type: 'alternative';
	requirements: Requirement[];
}

export interface ConcatenationRequirement extends BaseRequirement {
	type: 'concatenation';
	requirements: Requirement[];
}

export function isRequirement(requirement: Requirement | null): requirement is Requirement {
	return requirement !== null;
}

// Factories

export function singleRequirementFactory(message: string): RequirementFactory<SingleRequirement> {
	return (path, value) => ({
		path,
		value,
		message,
		type: 'single',
	});
}

export function dependenceRequirementFactory(
	message: string,
	requirement: Requirement
): RequirementFactory<DependenceRequirement> {
	return (path, value) => ({
		path,
		value,
		requirement,
		message,
		type: 'dependence',
	});
}

export function alternativeRequirementFactory(
	message: string,
	requirements: Requirement[]
): RequirementFactory<AlternativeRequirement | DependenceRequirement> {
	if (requirements.length < 1) {
		throw new Error('There must be 1 or more requirements');
	}

	if (requirements.length === 1) {
		return dependenceRequirementFactory(message, requirements[0]);
	}

	return (path, value) => ({
		path,
		value,
		message,
		type: 'alternative',
		requirements: requirements.reduce(
			(res, requirement) =>
				// (a || (b || (c && d)) || e) === (a || b || (c && d) || e)
				requirement.type === 'alternative' ? [...res, ...requirement.requirements] : [...res, requirement],
			[] as Requirement[]
		),
	});
}

export function concatenationRequirementFactory(
	message: string,
	requirements: Requirement[]
): RequirementFactory<ConcatenationRequirement | DependenceRequirement> {
	if (requirements.length < 1) {
		throw new Error('There must be 1 or more requirements');
	}

	if (requirements.length === 1) {
		return dependenceRequirementFactory(message, requirements[0]);
	}

	return (path, value) => ({
		path,
		value,
		message,
		type: 'concatenation',
		requirements: requirements.reduce(
			(res, requirement) =>
				// (a && (b && (c || d)) && e) === (a && b && (c || d) && e)
				requirement.type === 'concatenation' ? [...res, ...requirement.requirements] : [...res, requirement],
			[] as Requirement[]
		),
	});
}

// examples:
// const numberRequirementFactory = singleRequirementFactory('Value should be a number');

// function equalToRequirementFactory(validValue: any): RequirementFactory<SingleRequirement> {
// 	return singleRequirementFactory(`Value should equal to ${validValue}`);
// }

// function optionalRequirementFactory(requirement: Requirement): RequirementFactory<AlternativeRequirement> {
// 	const undefinedFactory = equalToRequirementFactory(undefined);

// 	return (path, value) => alternativeRequirementFactory('', [undefinedFactory(path, value), requirement])(path, value);
// }

// const optionalNumberRequirementFactory: RequirementFactory<AlternativeRequirement> = (path, value) =>
// 	optionalRequirementFactory(numberRequirementFactory(path, value))(path, value);

// function someRequirementFactory(requirement: Requirement): RequirementFactory<DependenceRequirement> {
// 	return dependenceRequirementFactory('Some item should follow the requirement', requirement);
// }
