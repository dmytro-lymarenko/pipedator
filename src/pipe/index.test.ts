import { createValidator, ValidationContext } from '../core';
import { singleRequirementFactory } from '../core/requirements';
import { pipe } from './index';

describe('pipe()', () => {
	it('should always succeed when validators are not passed down', () => {
		expect(pipe([]).validate(1)).toBeNull();
	});

	it('should invoke passed validator', () => {
		const validate = jest.fn(() => null);

		pipe([createValidator({ validate })]).validate(1);

		expect(validate.mock.calls.length).toBe(1);
		expect(validate.mock.calls[0][0]).toBe(1);
	});

	it('should invoke all validators when they all succeed', () => {
		const callbacks = [jest.fn(() => null), jest.fn(() => null), jest.fn(() => null)];

		expect(pipe(callbacks.map(callback => createValidator({ validate: callback }))).validate(1)).toBeNull();

		callbacks.forEach(callback => {
			expect(callback.mock.calls.length).toBe(1);
		});
	});

	it('should return the first met error in pipe chain of validators', () => {
		const callbacks = [
			jest.fn(() => null),
			jest.fn((value: any, ctx: ValidationContext) => singleRequirementFactory('Test')(ctx.path, value)),
			jest.fn(() => null),
		];

		expect(pipe(callbacks.map(callback => createValidator({ validate: callback }))).validate(1)).toMatchSnapshot();

		expect(callbacks[0].mock.calls.length).toBe(1);
		expect(callbacks[1].mock.calls.length).toBe(1);
		expect(callbacks[2].mock.calls.length).toBe(0);
	});

	it('should support custom message', () => {
		const callbacks = [
			jest.fn(() => null),
			jest.fn((value: any, ctx: ValidationContext) => singleRequirementFactory('Test')(ctx.path, value)),
		];

		expect(
			pipe(
				callbacks.map(callback => createValidator({ validate: callback })),
				'Custom message'
			).validate(1)
		).toMatchSnapshot();
	});
});
