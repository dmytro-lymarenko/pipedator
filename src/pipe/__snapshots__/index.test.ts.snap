// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pipe() should return the first met error in pipe chain of validators 1`] = `
Object {
  "children": null,
  "message": "Test",
  "path": "",
}
`;

exports[`pipe() should support custom message 1`] = `
Object {
  "children": null,
  "message": "Custom message",
  "path": "",
}
`;
